controllers/ocirepository_controller.go

```diff
@@ -942,3 +980,73 @@ func (r *OCIRepositoryReconciler) notify(ctx context.Context,
 		}
 	}
 }
+
+// Verify verifies the authenticity of the given image reference url. First, it tries to keyful approach by looking at
+// whether the given secret exists. Then, if it does not exist, it pushes a keyless approach for verification.
+func (r *OCIRepositoryReconciler) verify(ctx context.Context, obj *sourcev1.OCIRepository, url string) (bool, error) {
+	// get the public keys from the given secret
+	secretRef := obj.Spec.Verify.SecretRef
+
+	// Generate the registry credential keychain either from static credentials or using cloud OIDC
+	keychain, err := r.keychain(ctx, obj)
+	if err != nil {
+		return false, err
+	}
+	authnKeychain := signature.WithAuthnKeychain(keychain)
+
+	ref, err := name.ParseReference(url)
+	if err != nil {
+		return false, err
+	}
+
+	if secretRef != nil {
+		certSecretName := types.NamespacedName{
+			Namespace: obj.Namespace,
+			Name:      secretRef.Name,
+		}
+
+		var pubSecret corev1.Secret
+		if err := r.Get(ctx, certSecretName, &pubSecret); err != nil {
+			return false, err
+		}
+
+		// traverse all public keys and try to verify the signature
+		// this is brute-force approach, but it is ok for now
+		for k, data := range pubSecret.Data {
+			// search for public keys in the secret
+			if strings.HasSuffix(k, ".pub") {
+				verifier, err := signature.New(signature.WithPublicKey(data), authnKeychain)
+				if err != nil {
+					return false, err
+				}
+
+				signatures, _, err := verifier.VerifyImageSignatures(ctx, ref)
+				if err != nil {
+					r.eventLogf(ctx, obj, corev1.EventTypeNormal,
+						sourcev1.SourceVerifiedFailedReason,
+						"failed to verify image %s signature with key %s: %s", url, k, err)
```

Comment:
This will result in emitting kubernetes and notification-controller events for every failure.
We usually try to avoid spamming notification-controller with such events and use `events.EventTypeTrace` instead of `corev1.EventTypeNormal` if there's a necessity to log and record a kubernetes event about the failure.
But I'm not sure if there's any benefit in doing this here. Maybe a simple log is good enough without an event.

Also, considering that it's looping over the secret data, if the data contains some failure cases, even if verification succeeds, the logs would be filled with failure messages.

Reply:
> Maybe a simple log is good enough without an event.

There is no necessity to log and record a Kubernetes event here. Thanks for the suggestion, @darkowlzz, I've changed to code to just log this with the following code:

```go
ctrl.LoggerFrom(ctx).Error(err, "failed to verify image %s signature with key %s", url, k)
```

Reply:
Even with a single log, it's not considering my other concern

> Also, considering that it's looping over the secret data, if the data contains some failure cases, even if verification succeeds eventually, the logs would be filled with failure messages.

Maybe instead of logging immediately in each iteration, collect the failures and log only if no key was found in the secret. That way, we don't log any failure cases if eventually it succeeds due to multiple keys in the secret.
Also, it may be generally better to reduce the number of failure logs. Instead of 10 different failure logs, just one log saying none of the keys in the secret succeeded to verify.

Reply:
Done! Can you please check again? https://github.com/fluxcd/source-controller/pull/890